### PR TITLE
fips: update DSA security check to fix legacy verify strengths

### LINKED ISSUE #21185: Incorrect DSA verification check in FIPS provider
From openssl-users:

> Our organization signs image artifacts with 2048-bit DSA keys before releasing them to the field. Some of these signatures fail to verify when using the OpenSSL 3.0 FIPS provider. It turns out that while most of our signing keys are (L,N)=(2048,256), two early keys created long ago are (2048,160) and the signatures that fail to verify were created with these keys. Disabling security checks in the configuration file resolves this but I'd prefer not do that and inadvertently let something else non-compliant go undetected.
> 
> I discovered this code in providers/common/securitycheck.c:
> 
> ```c
>         /*
>          * For Digital signature verification DSA keys with < 112 bits of
>          * security strength (i.e L < 2048 bits), are still allowed for legacy
>          * use. The bounds given in SP800 131Ar2 - Table 2 are
>          * (512 <= L < 2048 and 160 <= N < 224)
>          */
>         if (!sign && L < 2048)
>             return (L >= 512 && N >= 160 && N < 224);
> ```
> 
> I am by no means an expert in cryptography but this logic does not seem to match my interpretation of the spec which for legacy use allows:
> 
> ```c
>     ((512 <= L < 2048) or (160 <= N < 224))
> ```
> 
> with "or" being the operative word here. OpenSSL is making this an "and" condition. Doesn't 800-131Ar2 allow (2048,160) when verifying a DSA signature, or am I misreading the spec?

### PR DESCRIPTION
**Pull request by paulidale**
3.0.x will need a change to the fips version line in the test (to >3.0.10)

Refer SP 800-131Ar2 table 2:
    https://csrc.nist.gov/publications/detail/sp/800-131a/rev-2/final

Fixes #21185

- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Looks like you have some test failures..

**Comment by paulidale** [comment]
Test failures should be addressed, not in an ideal way.

I don't seem to be able to generate a key from the command line that can't be verified by the FIPS provider.

File: providers/common/securitycheck.c
Comment (slontis) [inline providers/common/securitycheck.c:175]: L=256?

Comment (paulidale) [reply providers/common/securitycheck.c:175]: I think it would be from a strict reading of the standard which says: `((512 ≤ L < 2048) or (160 ≤ N < 224))`
Satisfying _either_ side makes the _or_ true.  So, 512 ≤ L < 2048 is acceptable regardless of N as is 160 ≤ N < 224 regardless of L.  I think that's stupid so the code is a bit stricter.

Comment (slontis) [reply providers/common/securitycheck.c:175]: Looks like that check has been there since at least 1.1.1

Comment (paulidale) [reply providers/common/securitycheck.c:175]: Yeah, that's relevant too.  Fix pushed.

I suspect that the 131Ar2 standard is being silly.  Rather than spelling out legal cases, it just has a blanket statement that's overly broad.

Comment (paulidale) [reply providers/common/securitycheck.c:175]: Does this mean you think we should be checking N = 160, 224 and 256 only despite what 131Ar2 says?

Comment (slontis) [reply providers/common/securitycheck.c:175]: Maybe OTC decision? I dont mind either way. It has been like this is openSSL for quite some time..

Comment (paulidale) [reply providers/common/securitycheck.c:175]: Or just leave the fix with what I did first?  That fixes the original issue just fine.
I doubt anyone is broken by the stricter check in `dsa_do_verify()` given how long it's been there.

File: crypto/dsa/dsa_ossl.c
Comment (slontis) [inline crypto/dsa/dsa_ossl.c:376]: Isnt it 160 =< N <= 224 or N == 256?

File: providers/common/securitycheck.c
Comment (slontis) [reply providers/common/securitycheck.c:175]: yeah probably - I obviously missed it when doing the check..

File: providers/common/securitycheck.c
Comment (levitte) [inline providers/common/securitycheck.c:182]: (L ≥ 512 and N ≥ 160 and (L < 2048 or N < 224))

That's how I read this.  Correct?  It might be good to spell this out in the comment, to make it easier to any reader to compare to the legacy logic in SP 800-131Ar2.

File: crypto/dsa/dsa_ossl.c
Comment (paulidale) [reply crypto/dsa/dsa_ossl.c:376]: They are the allowed lengths.  This is the opposite check for not allowed lengths.
I.e. the following are not allowed:

- N < 160
- 224 < N < 256
- N > 256

which is what's being checked.

Comment (paulidale) [reply crypto/dsa/dsa_ossl.c:376]: Kind of moot if this change is dropped....

File: providers/common/securitycheck.c
Comment (paulidale) [reply providers/common/securitycheck.c:182]: Did you look at 131Ar2? :grin:

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
> 3.0.x will need a change to the fips version line in the test (to >3.0.10)

That's questionable - Don't you with the fips provider compat tests test 3.0 tip of the branch with for example 3.1.1 (as the validated FIPS module?) -> it would fail that test then.

Otherwise, keeping the FIPSversion check as is IMO just means the legacy key verification won't be tested which is IMO acceptable.

**Comment by paulidale** [comment]
Happy to omit the test on back port, that's reasonable.

**Comment by paulidale** [comment]
Dropping commit 5b81e5f446fe5156f72ea4407bf5cdbdd1c0959e since that restriction has been there forever and doesn't impact the reported issue.

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
@openssl/committers second review?

**Review by slontis — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the reviews.

**Pull request closed without merging**
